File: packages/kotlin/src/main/kotlin/com/ecollect/sdk/ui/EcollectCardField.kt
package com.ecollect.sdk.ui

/**
 * Placeholder for Android View / Jetpack Composable for secure card capture.
 *
 * In a real Android project this would extend android.view.View (or be a @Composable).
 * This file shows the intended interface for the UI component.
 *
 * Security model:
 *  - Card data (PAN, CVV, expiry) is captured in this isolated view
 *  - Data is sent DIRECTLY to the ecollect tokenization endpoint
 *  - The merchant app NEVER sees the raw PAN; it only receives a TokenId
 *
 * Usage:
 * ```kotlin
 * val cardField = EcollectCardField()
 * cardField.setSessionToken(sessionToken)
 * cardField.setEntityCode(entityCode)
 * cardField.setOnTokenReceived { tokenId ->
 *     // Send tokenId to your backend to complete the payment
 *     backend.processPayment(tokenId)
 * }
 * ```
 */
class EcollectCardField {
    private var onTokenReceived: ((tokenId: String) -> Unit)? = null
    private var sessionToken: String? = null
    private var entityCode: Int? = null
    private var paymentSystem: String? = null
    private var fiCode: String? = null

    /**
     * Set callback to receive the TokenId once tokenization is complete.
     * The merchant app should send this TokenId to their backend.
     */
    fun setOnTokenReceived(callback: (tokenId: String) -> Unit) {
        onTokenReceived = callback
    }

    /**
     * Set the session token (obtained from the merchant's backend).
     * The ApiKey should NEVER be stored on the device or passed here.
     */
    fun setSessionToken(sessionToken: String) {
        this.sessionToken = sessionToken
    }

    /**
     * Set the entity (merchant) code.
     */
    fun setEntityCode(entityCode: Int) {
        this.entityCode = entityCode
    }

    /**
     * Set the payment system code.
     */
    fun setPaymentSystem(paymentSystem: String) {
        this.paymentSystem = paymentSystem
    }

    /**
     * Set the financial institution code (e.g., VISA, MASTERCARD).
     */
    fun setFiCode(fiCode: String) {
        this.fiCode = fiCode
    }

    // In a real Android View, this would be triggered by user interaction
    // (e.g., a Submit button tap) and would call the ecollect tokenCommand API
    // with Command=SAVE or Command=GET, then invoke onTokenReceived with the result.
}

File: packages/swift/Sources/EcollectSDK/UI/EcollectCardField.swift
import Foundation

/// Placeholder UI component for secure card data capture.
///
/// In a real iOS project this would extend UIView (UIKit) or implement SwiftUI's View protocol.
/// The component renders an isolated input view where the cardholder enters their PAN, CVV,
/// and expiration date. The data goes directly from the component to ecollect's tokenization
/// endpoint — the PAN never reaches the merchant's app code or backend.
///
/// Usage (conceptual):
/// ```swift
/// let cardField = EcollectCardField()
/// cardField.sessionToken = sessionToken
/// cardField.entityCode = 123
/// cardField.onTokenReceived = { tokenId in
///     // Pass tokenId to your backend to complete the payment
/// }
/// try await cardField.submit()
/// ```
public final class EcollectCardField {

    // MARK: - Configuration

    /// Session token obtained from the merchant's backend. Required before calling `submit()`.
    public var sessionToken: String = ""

    /// Merchant entity code assigned by ecollect.
    public var entityCode: Int = 0

    // MARK: - Callbacks

    /// Called when ecollect returns a TokenId after successful card capture.
    public var onTokenReceived: ((String) -> Void)?

    /// Called when an error occurs during tokenization.
    public var onError: ((EcollectError) -> Void)?

    public init() {}

    // MARK: - Submit

    /// Submits the captured card data directly to ecollect for tokenization.
    /// Returns the TokenId on success.
    ///
    /// - Returns: The ecollect TokenId.
    /// - Throws: `EcollectError` on failure.
    ///
    /// Note: This is a stub. In a real implementation this method would:
    /// 1. Read card data from the secure isolated input fields.
    /// 2. Validate client-side (Luhn, expiry).
    /// 3. POST directly to ecollect's tokenCommand endpoint.
    /// 4. Return the TokenId (never the raw PAN).
    public func submit() async throws -> String {
        // Stub — replace with real implementation in a UIKit/SwiftUI context.
        guard !sessionToken.isEmpty else {
            throw EcollectError.invalidConfig("sessionToken must be set before calling submit()")
        }
        guard entityCode > 0 else {
            throw EcollectError.invalidConfig("entityCode must be set before calling submit()")
        }
        // Real implementation: capture card fields, validate, tokenize via EcollectSDK.
        return ""
    }
}
